x/airdrop/types/msgs.go

```diff
@@ -0,0 +1,535 @@
+	}
+
+	if !msg.DistributionEndDate.After(*msg.DistributionStartDate) {
+		return errors.New("distribution end date must be after the start date")
+	}
+	if !msg.ClaimTypeDeadlineDate.After(*msg.DistributionStartDate) {
+		return errors.New("claim type deadline date must be after the distribution start date")
+	}
+	if !msg.ClaimTypeDeadlineDate.Before(*msg.DistributionEndDate) {
+		return errors.New("claim type deadline date must be before the distribution end date")
+	}
+	if !msg.ClawbackDate.After(*msg.DistributionEndDate) {
+		return errors.New("clawback date must be after the distribution end date")
+	}
+
+	if msg.EarlyClaimPenalty.IsNil() {
+		return errors.New("early claim penalty must be specified")
+	}
+
+	if msg.EarlyClaimPenalty.LT(sdk.ZeroDec()) || msg.EarlyClaimPenalty.GT(sdk.OneDec()) {
+		return errors.New("early claim penalty must be between 0 and 1")
+	}
+
+	if _, err := sdk.AccAddressFromBech32(msg.DistributionAddress); err != nil {
+		return errorsmod.Wrapf(sdkerrors.ErrInvalidAddress, "invalid distribution address (%s)", err)
+	}
+
+	return nil
+}
+
+// ----------------------------------------------
+//               MsgUpdateAirdrop
+// ----------------------------------------------
+
+func NewMsgUpdateAirdrop(
+	admin string,
+	airdropId string,
+	rewardDenom string,
+	distributionStartDate *time.Time,
+	distributionEndDate *time.Time,
+	clawbackDate *time.Time,
+	claimDeadlineDate *time.Time,
+	earlyClaimPenalty sdk.Dec,
+	distributionAddress string,
+) *MsgUpdateAirdrop {
+	return &MsgUpdateAirdrop{
+		Admin:                 admin,
+		AirdropId:             airdropId,
+		RewardDenom:           rewardDenom,
+		DistributionStartDate: distributionStartDate,
+		DistributionEndDate:   distributionEndDate,
+		ClawbackDate:          clawbackDate,
+		ClaimTypeDeadlineDate: claimDeadlineDate,
+		EarlyClaimPenalty:     earlyClaimPenalty,
+		DistributionAddress:   distributionAddress,
+	}
+}
+
+func (msg MsgUpdateAirdrop) Type() string {
+	return TypeMsgUpdateAirdrop
+}
+
+func (msg MsgUpdateAirdrop) Route() string {
+	return RouterKey
+}
+
+func (msg *MsgUpdateAirdrop) GetSigners() []sdk.AccAddress {
+	claimer, err := sdk.AccAddressFromBech32(msg.Admin)
+	if err != nil {
+		panic(err)
+	}
+	return []sdk.AccAddress{claimer}
+}
+
+func (msg *MsgUpdateAirdrop) GetSignBytes() []byte {
+	bz := ModuleCdc.MustMarshalJSON(msg)
+	return sdk.MustSortJSON(bz)
+}
+
+func (msg *MsgUpdateAirdrop) ValidateBasic() error {
+	if _, err := sdk.AccAddressFromBech32(msg.Admin); err != nil {
+		return errorsmod.Wrapf(sdkerrors.ErrInvalidAddress, "invalid address (%s)", err)
+	}
+	if err := utils.ValidateAdminAddress(msg.Admin); err != nil {
+		return err
+	}
+
+	if msg.AirdropId == "" {
+		return errors.New("airdrop-id must be specified")
+	}
+	if msg.RewardDenom == "" {
+		return errors.New("reward denom must be specified")
+	}
+
+	if msg.DistributionStartDate == nil || *msg.DistributionStartDate == (time.Time{}) {
```

Comment:
Very good coverage in the `ValidateBasic`s! 

nit, should we condense the validation in `MsgUpdateAirdrop` and `MsgCreateAirdrop`?

Reply:
definitely! will update

Reply:
done

Reply:
Amazing, resolving!